### PR TITLE
[Habits view] The user is able to select the duration of habit formation less than the primary duration #5480

### PR DESCRIPTION
**Pull request by ABbondar**
# GreenCity PR
https://github.com/ita-social-projects/GreenCity/issues/5480

## Summary Of Changes :fire:
Changed method updateShoppingItemList to updateHabitAssignDuration in HabitAssignController

## Issue Link :clipboard:
_&lt;Link to the issue&gt;_

## Added
Validation for the duration of habit formation
Tests for a new methods

## Changed
* _&lt;detail item of what was changed&gt;_
* _&lt;describe functionality that was changed&gt;_

## Deleted
Unused methods, dto's and tests

## How to test :clipboard:
It can be test via swagger in habit-assign-controller, endpoint [/habit/assign/{habitAssignId}/update-habit-duration]

# PR Checklist Forms

_(to be filled out by PR submitter)_
- [ ] Code is up-to-date with the `dev` branch.
- [x] You've successfully built and run the tests locally.
- [x] There are new or updated unit tests validating the change.
- [x] JIRA/ Github Issue number & title in PR title (ISSUE-XXXX: Ticket title)
- [x] This template filled (above this section).
- [x] Sonar's report does not contain bugs, vulnerabilities, security issues, code smells ar duplication
- [x] `NEED_REVIEW` and `READY_FOR_REVIEW` labels are added.
- [x] All files reviewed before sending to reviewers

### CODE REVIEW

File: core/src/main/java/greencity/controller/HabitAssignController.java [hunk 1]
```diff
@@ -77,27 +91,27 @@ public ResponseEntity<HabitAssignManagementDto> assignCustom(@PathVariable Long
...
-     * @param userVO                   {@link UserVO} instance.
-     * @param habitAssignPropertiesDto {@link HabitAssignPropertiesDto} instance.
+     * @param habitAssignId {@link HabitVO} id.
+     * @param userVO        {@link UserVO} instance.
+     * @param duration      {@link Integer} with needed duration.
      * @return {@link ResponseEntity}.
      */
-    @ApiOperation(value = "Update user shopping item list and habit assign duration.")
+    @PutMapping("/{habitAssignId}/update-habit-duration")
+    @ApiOperation(value = "Update duration of habit assigned for user.")
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/HabitAssignController.java]: please add that habit with `habitAssignId` will be updated

File: core/src/test/java/greencity/controller/HabitAssignControllerTest.java [hunk 2]
```diff
@@ -90,15 +90,11 @@ void updateAssignByHabitId() throws Exception {
...
-        Gson gson = new Gson();
-        String json = gson.toJson(propertiesDto);
-        mockMvc.perform(put(habitLink + "/{habitId}/update-user-shopping-item-list", 1L)
-            .content(json)
+    void updateHabitAssignDurationTest() throws Exception {
+        mockMvc.perform(put(habitLink + "/{habitId}/update-habit-duration?duration=15", 1L)
             .contentType(MediaType.APPLICATION_JSON))
             .andExpect(status().isOk());
-        verify(habitAssignService).updateUserShoppingItemListAndDuration(1L, null, propertiesDto);
+        verify(habitAssignService).updateUserHabitInfoDuration(1L, null, 15);
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/HabitAssignControllerTest.java:97]: please add tests to test `AppConstant.MIN_DAYS_DURATION` and `AppConstant.MAX_DAYS_DURATION`

File: core/src/test/java/greencity/controller/HabitAssignControllerTest.java [hunk 3]
```diff
@@ -90,15 +90,11 @@ void updateAssignByHabitId() throws Exception {
...
 
     @Test
-    void updateShoppingItemList() throws Exception {
-        HabitAssignPropertiesDto propertiesDto = ModelUtils.getHabitAssignPropertiesDto();
-        Gson gson = new Gson();
-        String json = gson.toJson(propertiesDto);
-        mockMvc.perform(put(habitLink + "/{habitId}/update-user-shopping-item-list", 1L)
-            .content(json)
+    void updateHabitAssignDurationTest() throws Exception {
+        mockMvc.perform(put(habitLink + "/{habitId}/update-habit-duration?duration=15", 1L)
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/HabitAssignControllerTest.java]: please change `habitId`  to `habitAssignId`

File: dao/src/main/java/greencity/repository/HabitAssignRepo.java [hunk 4]
```diff
@@ -85,6 +85,20 @@ Optional<HabitAssign> findByHabitIdAndUserIdAndCreateDate(@Param("habitId") Long
...
+    /**
+     * Method to find {@link HabitAssign}'s by {@link User} and {@link HabitAssign}
+     * id and INPROGRESS status.
+     *
+     * @param habitAssignId {@link Long} id.
+     * @param userId        {@link Long} id.
+     * @return {@link HabitAssign} instance.
+     * @author Anton Bondar
+     */
+    @Query(value = "SELECT DISTINCT ha FROM HabitAssign ha"
```
Comment (RomaMocherniuk) [inline dao/src/main/java/greencity/repository/HabitAssignRepo.java]: Why do we need `DISTINCT` here?
can we have several HabitAssign with the same `habitAssignId` and `userId`?

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 5]
```diff
@@ -231,40 +230,27 @@ public Integer getReadinessPercent(HabitAssignDto habitAssign) {
...
-                shoppingListItemRepo.getShoppingListByListOfId(habitAssignPropertiesDto
-                    .getDefaultShoppingListItems());
-            List<UserShoppingListItem> userShoppingListItems = shoppingListItems.stream()
-                .map(s -> buildUserShoppingListItems(s, habitAssign))
-                .collect(Collectors.toList());
-            Hibernate.initialize(habitAssign.getUserShoppingListItems());
-            userShoppingListItemRepo.deleteAll(habitAssign.getUserShoppingListItems());
-            habitAssign.setUserShoppingListItems(userShoppingListItems);
+        if (duration < habitAssign.getWorkingDays()) {
+            throw new BadRequestException(ErrorMessage.INVALID_DURATION);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java:249]: should we check if `duration > habitAssign.getWorkingDays()`?

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 6]
```diff
@@ -231,40 +230,27 @@ public Integer getReadinessPercent(HabitAssignDto habitAssign) {
...
-        HabitAssignPropertiesDto habitAssignPropertiesDto) {
-        if (!habitRepo.existsById(habitId)) {
-            throw new NotFoundException(ErrorMessage.HABIT_NOT_FOUND_BY_ID + habitId);
+    public HabitAssignUserDurationDto updateUserHabitInfoDuration(Long habitAssignId, Long userId, Integer duration) {
+        if (!habitAssignRepo.existsById(habitAssignId)) {
+            throw new NotFoundException(ErrorMessage.HABIT_NOT_FOUND_BY_ID + habitAssignId);
         }
-        HabitAssign habitAssign = habitAssignRepo.findByHabitIdAndUserIdAndStatusIsInprogress(habitId, userId)
+        HabitAssign habitAssign = habitAssignRepo.findByHabitAssignIdUserIdAndStatusIsInProgress(habitAssignId, userId)
             .orElseThrow(() -> new InvalidStatusException(ErrorMessage.HABIT_ASSIGN_STATUS_IS_NOT_INPROGRESS));
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java]: `habitAssignRepo.findByHabitAssignIdUserIdAndStatusIsInProgress` will also return `Optional.empty` for `INPROGRESS` habitAssign if ` ha.user.id != :userId`
See https://github.com/ita-social-projects/GreenCity/pull/5531/files#diff-05169418d0f08eeb9efc32c9ae364ca45acf5015e13946bd7fffc632a1546bdcR98
Should we use another exception message in this case?

File: core/src/test/java/greencity/controller/HabitAssignControllerTest.java (hunk 3 above)
Comment (ABbondar) [reply core/src/test/java/greencity/controller/HabitAssignControllerTest.java]: changed

**Review by RomaMocherniuk — APPROVED** [review]

**Review by juseti — APPROVED** [review]

**Review by Markiro1 — APPROVED** [review]

**Review by sergoliarnik — APPROVED** [review]

**Pull request merged**